--!strict

export type CollisionGroupDefinition = {
	name: string,
	collidesWith: {string},
}

local PLAYERS_COLLISION_GROUP: string = "PLAYERS_COLLISION_GROUP"
local BOUNDS_COLLISION_GROUP: string = "BOUNDS_COLLISION_GROUP"
local BOUNDS_QUERY_GROUP: string = "BOUNDS_QUERY_GROUP"
local HURTBOX_COLLISION_GROUP: string = "HURTBOX_COLLISION_GROUP"
local TRIGGERBOX_COLLISION_GROUP: string = "TRIGGERBOX_COLLISION_GROUP"
local TRIGGERBOX_QUERY_GROUP: string = "TRIGGERBOX_QUERY_GROUP"
local DEFAULT_COLLISION_GROUP: string = "Default"

local COLLISION_GROUP_DEFINITIONS: { CollisionGroupDefinition } = {
	{
		name = PLAYERS_COLLISION_GROUP,
		collidesWith = {
			DEFAULT_COLLISION_GROUP,
		},
	},
	{
		name = BOUNDS_COLLISION_GROUP,
		collidesWith = {},
	},
	{
		name = BOUNDS_QUERY_GROUP,
		collidesWith = {
			BOUNDS_COLLISION_GROUP,
		},
	},
	{
		name = HURTBOX_COLLISION_GROUP,
		collidesWith = {},
	},
	{
		name = TRIGGERBOX_COLLISION_GROUP,
		collidesWith = {},
	},
	{
		name = TRIGGERBOX_QUERY_GROUP,
		collidesWith = {
			TRIGGERBOX_COLLISION_GROUP,
		},
	}
}

return {
	PLAYERS_COLLISION_GROUP = PLAYERS_COLLISION_GROUP,
	BOUNDS_COLLISION_GROUP = BOUNDS_COLLISION_GROUP,
	BOUNDS_QUERY_GROUP = BOUNDS_QUERY_GROUP,
	HURTBOX_COLLISION_GROUP = HURTBOX_COLLISION_GROUP,
	DEFAULT_COLLISION_GROUP = DEFAULT_COLLISION_GROUP,
	TRIGGERBOX_COLLISION_GROUP = TRIGGERBOX_COLLISION_GROUP,
	TRIGGERBOX_QUERY_GROUP = TRIGGERBOX_QUERY_GROUP,
	COLLISION_GROUP_DEFINITIONS = COLLISION_GROUP_DEFINITIONS,
}
